packages/arb-ts/src/lib/L1ToL2Message/L1ToL2Message.ts

```diff
@@ -31,86 +29,123 @@ export interface L1ToL2MessageReceipt {
+ * type L1ToL2MessageWriter.
+ */
+type L1ToL2MessageReaderOrWriter<T extends Provider | Signer> =
+  T extends Provider ? L1ToL2MessageReader : L1ToL2MessageWriter
 
-  constructor(
-    signersAndProviders: SignersAndProviders,
-    public readonly l2TicketCreationTxnHash: string,
-    public readonly messageNumber?: BigNumber,
-    l1TxnHash?: string
-  ) {
-    super()
-    this.initSignersAndProviders(signersAndProviders)
-    this.arbRetryableActions = new RetryableActions(signersAndProviders)
-    this.l1TxnHash = l1TxnHash
-  }
-  static async initFromL1Txn(
-    signersAndProviders: SignersAndProviders,
-    l1Txn: string | TransactionReceipt,
+export class L1ToL2Message {
+  protected static getMessageOrThrow(
+    l1TxnReceipt: TransactionReceipt,
     messageNumberIndex?: number
-  ): Promise<L1ToL2Message> {
-    const l1TxnHash = typeof l1Txn === 'string' ? l1Txn : l1Txn.transactionHash
-    const allL1ToL2Messages = await L1ToL2Message.initAllFromL1Txn(
-      signersAndProviders,
-      l1Txn
-    )
-    const messageCount = allL1ToL2Messages.length
-    if (!messageCount)
-      throw new Error(`No l1 to L2 message found for ${l1TxnHash}`)
+  ) {
+    const messageNumbers = getMessageNumbersFromL1TxnReceipt(l1TxnReceipt)
+    if (messageNumbers === undefined)
+      throw new Error(
+        `No l1 to L2 message found for ${l1TxnReceipt.transactionHash}`
+      )
 
-    if (messageNumberIndex !== undefined && messageNumberIndex >= messageCount)
+    if (
+      messageNumberIndex !== undefined &&
+      messageNumberIndex > messageNumbers.length
+    )
       throw new Error(
-        `Provided message number out of range for ${l1TxnHash}; index was ${messageNumberIndex}, but only ${messageCount} messages`
+        `Provided message number out of range for ${l1TxnReceipt.transactionHash}; index was ${messageNumberIndex}, but only ${messageNumbers.length} messages`
       )
-    if (messageNumberIndex === undefined && messageCount > 1)
+    if (messageNumberIndex === undefined && messageNumbers.length > 1)
       throw new Error(
-        `${messageCount} L2 messages for ${l1TxnHash}; must provide messageNumberIndex (or use (signersAndProviders, l1Txn))`
+        `${messageNumbers.length} L2 messages for ${l1TxnReceipt.transactionHash}; must provide messamessageNumberIndex (or use initAllFromL1Txn)`
       )
 
-    return allL1ToL2Messages[messageNumberIndex || 0]
+    return messageNumbers[messageNumberIndex || 0]
   }
 
-  static async initAllFromL1Txn(
-    signersAndProviders: SignersAndProviders,
-    l1Txn: string | TransactionReceipt
-  ): Promise<L1ToL2Message[]> {
-    const l1TxnReceipt = await getTxnReceipt(
-      l1Txn,
-      signersAndProviders.l1Provider
-    )
-    const l1TxnHash = l1TxnReceipt.transactionHash
+  public static async fromL1ReceiptAll<T extends Provider | Signer>(
+    l2SignerOrProvider: T,
+    l1TxnReceipt: TransactionReceipt
```

Comment:
We could have a "from L1 Hash" version of all of these that just wraps around them and uses getTxnReceipt w/ an L1 provider... maybe that's overkill?

Reply:
We'd still need 3 methods I think: fromL1Hash, fromL1HashAll and fromL2Ticket. I think getting the receipt for a hash is a very common thing that people using ethers will be useful, so I think I'd rather leave them as they are.